Add SensorTile.box board support and sample to test it

avisconti opened the pull request:
Add the BSP for SensorTile.box board with documentation.
And a sample code to test it out.

avisconti committed:
drivers: pinmux: stm32l4x: Add missing SPI macros

This commit adds:

 1. pin definitions for SPI1 on PE12-15 on STM32L4 devices
 2. SPI3 on PA15_SPI3_NSS

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
boards: arm: Add support for SensorTile.box board

The SensorTile.box is a board designed for IoT applications
embedding a wide range of intelligent low power MEMS sensors,
a STM32L4 microcontroller to manage sensor configuration and
process sensor output data, a micro-USB battery charging
interface and an ST Bluetooth Low Energy module for wireless
communication with a BLE-enabled smartphone.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
sample/board: add SensorTile.box sample for testing

Add sample to test SensorTile.box board.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

dbkinder commented on samples/boards/sensortile_box/README.rst:
```diff
@@ -0,0 +1,78 @@
+    :board: sensortile_box
+    :goals: build flash
+
+Please note that flashing the board requires few preliminary steps described
```
change to ``requires a few``

dbkinder commented on samples/boards/sensortile_box/README.rst:
```diff
@@ -0,0 +1,78 @@
+Please note that flashing the board requires few preliminary steps described
+in :ref:`sensortile_box`.
+
+Then, power cycle the board disconnecting and reconnecting the USB cable.
```
change to ``board by disconnecting``

dbkinder commented on samples/boards/sensortile_box/README.rst:
```diff
@@ -0,0 +1,78 @@
+in :ref:`sensortile_box`.
+
+Then, power cycle the board disconnecting and reconnecting the USB cable.
+Run your favorite terminal program to listen for output.
```
Since we won't hear the output, let's change ``listen for`` to ``display``

dbkinder commented on samples/boards/sensortile_box/README.rst:
```diff
@@ -0,0 +1,78 @@
+Sample Output
+=============
+
+The sample code outputs sensors data on SensorTile.box console.
```
change to ``on the``

dbkinder commented on samples/boards/sensortile_box/README.rst:
```diff
@@ -0,0 +1,78 @@
+    1:: lsm6dsox acc trig 836
+    1:: lsm6dsox gyr trig 836
+
+    <repeats endlessly every 2s>
```
May add a blank line after the code block and another paragraph:
```
If you move the board around or put your finger on the temperature
sensor, you will see the accelerometer, gyro, and temperature values change. 
```

avisconti commented on boards/arm/sensortile_box/doc/index.rst:
```diff
@@ -0,0 +1,204 @@
+Overview
+********
+
+The STEVAL-MKSBOX1V1 (SensorTile.box) is a ready-to-use box kit with wireless
```
ack

avisconti commented on boards/arm/sensortile_box/doc/index.rst:
```diff
@@ -0,0 +1,204 @@
+********
+
+The STEVAL-MKSBOX1V1 (SensorTile.box) is a ready-to-use box kit with wireless
+IoT and wearable sensor platform to help you use and develop apps based on
```
ok

avisconti commented on boards/arm/sensortile_box/doc/index.rst:
```diff
@@ -0,0 +1,204 @@
+------------
+
+- BUTTON = BOOT
+  (used to let the SensorTile.box entering DFU mode. See "Programming and Debugging"
```
ok

avisconti commented:
@dbkinder 
Thx! Applied all ur suggestions, re-based to latest master and re-pushed!

ydamigos commented on boards/arm/sensortile_box/sensortile_box.dts:
```diff
@@ -0,0 +1,182 @@
+};
+
+&spi2 {
+	status = "okay";
```
If you enable SPI2 in DT, better to do the same in board's Kconfig.defconfig.

avisconti replied:
@ydamigos
Actually I do not need spi2. Shall I remove everything or deleting just the 'status' row?

avisconti replied:
I removed everything

dbkinder commented on boards/arm/sensortile_box/doc/index.rst:
```diff
@@ -0,0 +1,206 @@
+------------
+
+- BUTTON = BOOT
+  (used to let the SensorTile.box entering DFU mode. See `Programming and Debugging`_
```
change ``entering`` to ``enter``

avisconti replied:
fixed thanks

dbkinder reviewed (changes requested):
one typo, otherwise LGTM

avisconti commented:
@dbkinder 
Fixed and rebased

avisconti commented:
I rebased this PR on top of #19068.

@dbkinder is now the doc fine?

@erwango @MaureenHelm 
Any thoughts on this PR?

erwango commented on boards/arm/sensortile_box/doc/index.rst:
```diff
@@ -0,0 +1,206 @@
+Programming and Debugging
+*************************
+
+There are 2 main entry points for flashing STM32FL4Rx SoCs, one using the ROM
```
STM32~~F~~L4Rx

erwango commented on boards/arm/sensortile_box/doc/index.rst:
```diff
@@ -0,0 +1,206 @@
+You can build and flash the provided sample application
+(:ref:`sensortile_box sensors`) that reads sensors data and outputs
+values on the console.
+
```
Since you provided the dfu-runner info, I guess that usual west flash command will work.
It would be good adding and example of west flash/debug commands for zephyr beginners.

erwango commented on boards/arm/sensortile_box/pinmux.c:
```diff
@@ -0,0 +1,68 @@
+	{STM32_PIN_PG8, STM32L4X_PINMUX_FUNC_PG8_I2C3_SDA},
+#endif /* CONFIG_I2C_3 */
+#ifdef CONFIG_SPI_1
+	{STM32_PIN_PE12, STM32L4X_PINMUX_FUNC_PE12_SPI1_NSS},
```
_NSS pins should be defined under 
```
#ifdef CONFIG_SPI_STM32_USE_HW_SS
```
Introduced in #17806

erwango commented on boards/arm/sensortile_box/pinmux.c:
```diff
@@ -0,0 +1,68 @@
+#endif /* CONFIG_SPI_1 */
+#ifdef CONFIG_SPI_2
+	/* SPI2 on the ST Morpho Connector CN10 pins 16, 30, 28, 26*/
+	{STM32_PIN_PB12, STM32L4X_PINMUX_FUNC_PB12_SPI2_NSS},
```
Ditto

erwango commented on boards/arm/sensortile_box/pinmux.c:
```diff
@@ -0,0 +1,68 @@
+#endif /* CONFIG_SPI_1 */
+#ifdef CONFIG_SPI_3
+	/* SPI3 is used in 3-wires mode */
+	{STM32_PIN_PA15, STM32L4X_PINMUX_FUNC_PA15_SPI3_NSS},
```
ditto

erwango commented on boards/arm/sensortile_box/sensortile_box_defconfig:
```diff
@@ -0,0 +1,62 @@
+CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y
+
+# Use HSI (16MHz) to feed into PLL
+#CONFIG_CLOCK_STM32_PLL_SRC_HSI=y
```
Comment not in line with configuration, or misleading.

erwango reviewed (changes requested):
Few comments otherwise is good and clean

dbkinder reviewed (approved):
LGTM, thanks for the doc updates!

avisconti commented on drivers/pinmux/stm32/pinmux_stm32l4x.h:
```diff
@@ -88,6 +88,8 @@
 #define STM32L4X_PINMUX_FUNC_PA15_USART2_RX                                    \
 	(STM32_PINMUX_ALT_FUNC_3 | STM32_PUPDR_NO_PULL)
+#define STM32L4X_PINMUX_FUNC_PA15_SPI3_NSS                                     \
+	(STM32_PINMUX_ALT_FUNC_5 | STM32_PUSHPULL_NOPULL)
```
Correct. Actually I reviewed the #19017 but somehow I missed to re-align my own pending PR.
Thanks!

erwango replied:
Nice catch, thanks @MaureenHelm !

avisconti commented on boards/shields/x_nucleo_iks01a2/x_nucleo_iks01a2.overlay:
```diff
@@ -27,7 +27,7 @@
 
 	lsm303agr-magn@1e {
-		compatible = "st,lis2mdl-magn","st,lsm303agr-magn";
+		compatible = "st,lis2mdl","st,lsm303agr-magn";
```
Yes, sure. I was waiting for the merging!

avisconti commented:
> @avisconti, no issue with adding STM32L4R9. It could actually be done in the current PR, which will be used as validation. From what I see STM32L4R9 is a direct superset of STM32L4R5, so addition should be minor:
> 
> * update soc/arm/st_stm32/stm32l4/Kconfig.soc with new reference
> * add soc/arm/st_stm32/stm32l4/Kconfig.defconfig.stm32l4r9xx file
> * add dts/arm/st/l4/stm32l4r9.dtsi as a superset of dts/arm/st/l4/stm32l4r5.dtsi
> * add dts/arm/st/l4/stm32l4r9X?.dtsi defining the soc memory attributes.

@erwango 
OK, next week I'll take a look at it. I may need your help though.

avisconti committed:
soc: arm: st: add stm32l4r9xx support

Add support for ST stm32l4r9xx System-on-Chip.

See http://www.st.com/en/microcontrollers/stm32l4r9-s9.html
for more details.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
> > @avisconti, no issue with adding STM32L4R9. It could actually be done in the current PR, which will be used as validation. From what I see STM32L4R9 is a direct superset of STM32L4R5, so addition should be minor:
> > 
> > * update soc/arm/st_stm32/stm32l4/Kconfig.soc with new reference
> > * add soc/arm/st_stm32/stm32l4/Kconfig.defconfig.stm32l4r9xx file
> > * add dts/arm/st/l4/stm32l4r9.dtsi as a superset of dts/arm/st/l4/stm32l4r5.dtsi
> > * add dts/arm/st/l4/stm32l4r9X?.dtsi defining the soc memory attributes.
> 
> @erwango
> OK, next week I'll take a look at it. I may need your help though.

@erwango,

I re-pushed it.
Now I added support for STM32L4R9 SoC. It is just an easy modification, so it would require some attention/suggestions from your side.

erwango commented on dts/arm/st/l4/stm32l4r9.dtsi:
```diff
@@ -0,0 +1,235 @@
+ */
+
+#include <mem.h>
+#include <st/l4/stm32l4.dtsi>
```
Why not including stm32l4r5.dtsi?
From what I could see l4r9 is a l4r5 + DSI, LTDC and GFX

avisconti replied:
> Why not including stm32l4r5.dtsi?
> From what I could see l4r9 is a l4r5 + DSI, LTDC and GFX

Seems quite good idea

avisconti replied:
@erwango 
Why not adding a stm32l4rx.dtsi equal to the current stm32l4rx.dtsi and then include it in both stm32l4r5.dtsi and stm32l4r9.dtsi ?

What do you think about that? Would that be more clear?

erwango replied:
@avisconti, we currently apply the same model for all stm32 series with a stm32xx.dtsi base, which is the "root" of the family (peripherals included in all variants of the series). Then this one is included the most simple SoCs variants which are then included in more complex SoCs and this goes on...
Your proposal is adding a second "root" to the stm32l4 series, which is a divergence from this model. Even if i seems fine, I'd prefer we stick to the initial scheme, so we don't see "root" files poping up in various places.

avisconti replied:
That's fine for me. I understand your point.

erwango commented on dts/arm/st/l4/stm32l4r9Xi.dtsi:
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright (c) 2018 Pushpal Sidhu
```
Use ST copyright

erwango commented on soc/arm/st_stm32/stm32l4/Kconfig.defconfig.stm32l4r9xx:
```diff
@@ -0,0 +1,34 @@
+# Kconfig - ST Microelectronics STM32L4R9xx MCU
+#
+# Copyright (c) 2018 Pushpal Sidhu
```
Use ST copyright

avisconti commented:
@erwango 
stm32l4r9.dtsi now only includes stm32l4r5.dtsi, that's all.
I think this was your intention.

Concerning flash size, is it same as stm32l4r5? I didn't check it to be honest...

erwango commented:
@avisconti, You're using "zi" ("Xi") variant so, it should be the same.

avisconti committed:
driver/sensor: IIS3DHHC: Fix DT_ macro typo for CS GPIO

The correct suffix name for selecting the GPIO for SPI Chip Select
is _CS_GPIOS_CONTROLLER and not _CS_GPIO_CONTROLLER.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
@dbkinder 

Since IIS3DHHC has been recently merged I decided to add to sample code also the test for it.
And I have changed the sample documentation.

@erwango @galak 

While testing IIS3DHHC for this board (is on SPI bus) I realize that there was some typos
on CS_GPIO DT name. So I fixed it and put the fix in this PR.
